### ISSUE TITLE
EarlyStopping Callback relative threshold mode

### ISSUE THREAD

**Issue by tlpss** [PyTorchLightning/pytorch-lightning#12094]
## 🚀 Feature

Provide the option to evaluate relative changes to the tracked metric in the `EarlyStoppingCallback` instead of the absolute difference.
### Motivation

I'm working on keypoint detection for robot manipulation and use a BCELoss in which the optimal value is the entropy of the ground truth heatmaps, which can change depending on the number of keypoints, number of channels, size of the gaussian blobs... Hence it is hard to specify the desired minimal absolute delta. It would be nice if we could specify a relative delta.

### Additional context

Inspired by the [Torch learning rate schedulers](https://pytorch.org/docs/stable/generated/torch.optim.lr_scheduler.ReduceLROnPlateau.html#torch.optim.lr_scheduler.ReduceLROnPlateau).

**Comment by carmocca** [comment]
@tlpss can you add more details about what sort of API would you find useful?

**Comment by tlpss** [comment]
@carmocca, I would propose to add an additional parameter `threshold_mode` that would interpret the delta value as either absolute or relative, is in Pytorch LR schedulers https://pytorch.org/docs/stable/generated/torch.optim.lr_scheduler.ReduceLROnPlateau.html

**Comment by awaelchli** [comment]
Adding another argument `threshold_mode` can probably done relatively easily as it looks like it will be orthogonal to the other ones. 
Alternatively, more broadly a LambdaEarlyStopping callback could probably be useful, where the user provides its own implementation in a function that gets the current value and the best and they return the decision whether to stop or continue.

**Comment by martinrohbeck** [comment]
Hey guys,
I was also looking for a EarlyStopping Callback based on relative changes in the loss metric. I implemented my own solution using `on_validation_epoch_end` now, but I think extending the `EarlyStopping` class [here](https://github.com/Lightning-AI/lightning/blob/3ac550f94b7fc245c42c165eb317e5f5e71e56ab/src/lightning/pytorch/callbacks/early_stopping.py#L34) with a new keyword (determining absolute or relative changes) makes a lot of sense.

What do you think? Is anybody already working on it?

Otherwise, I'm happy to give it a shot!

**Comment by BrunoBelucci** [comment]
Hello,
I also needed to use relative thresholds, but I think that the code from @tlpss is slightly wrong, especially if we want to be consistent with functions from pytorch that implement both modes ('abs' and 'rel'), as it is the case for example for `torch.optim.lr_scheduler.ReduceLROnPlateau`. If we are maximizing in 'rel' mode we are considering an improvement if 
$$\frac{current-best}{best}>\epsilon$$
therefore we should have
$$\frac{current}{(1+\epsilon)}>best$$
if we are minimizing we are considering an improvement if
$$\frac{best-current}{best}>\epsilon$$
therefore we should have
$$best>\frac{current}{(1-\epsilon)}$$

If the rest of the code stays the same with the definition of `self.monitor_op` and the change of sign of `self.min_delta`, in 'rel' mode we should use `if self.monitor_op(current / (1 + self.min_delta), self.best_score.to(current.device))`.

Here is my version of a modified EarlyStopping that implements both modes:
<pre><code>
class EarlyStopping_mod(EarlyStopping):
    def __init__(self, *args, threshold_mode='abs', **kwargs):
        self.threshold_mode = threshold_mode
        super().__init__(*args, **kwargs)

    def _evaluate_stopping_criteria(self, current: Tensor) -> Tuple[bool, Optional[str]]:
        should_stop = False
        reason = None
        if self.check_finite and not torch.isfinite(current):
            should_stop = True
            reason = (
                f"Monitored metric {self.monitor} = {current} is not finite."
                f" Previous best value was {self.best_score:.3f}. Signaling Trainer to stop."
            )
        elif self.stopping_threshold is not None and self.monitor_op(current, self.stopping_threshold):
            should_stop = True
            reason = (
                "Stopping threshold reached:"
                f" {self.monitor} = {current} {self.order_dict[self.mode]} {self.stopping_threshold}."
                " Signaling Trainer to stop."
            )
        elif self.divergence_threshold is not None and self.monitor_op(-current, -self.divergence_threshold):
            should_stop = True
            reason = (
                "Divergence threshold reached:"
                f" {self.monitor} = {current} {self.order_dict[self.mode]} {self.divergence_threshold}."
                " Signaling Trainer to stop."
            )
        <b>elif (self.threshold_mode == 'abs'
              and self.monitor_op(current - self.min_delta, self.best_score.to(current.device))):
            should_stop = False
            reason = self._improvement_message(current)
            self.best_score = current
            self.wait_count = 0
        elif (self.threshold_mode == 'rel'
              and self.monitor_op(current / (1 + self.min_delta), self.best_score.to(current.device))):
            should_stop = False
            reason = self._improvement_message(current)
            self.best_score = current
            self.wait_count = 0</b>
        else:
            self.wait_count += 1
            if self.wait_count >= self.patience:
                should_stop = True
                reason = (
                    f"Monitored metric {self.monitor} did not improve in the last {self.wait_count} records."
                    f" Best score: {self.best_score:.3f}. Signaling Trainer to stop."
                )

        return should_stop, reason

    def _improvement_message(self, current: Tensor) -> str:
        """Formats a log message that informs the user about an improvement in the monitored score."""
        if torch.isfinite(self.best_score):
            <b>if self.threshold_mode == 'abs':
                msg = (
                    f"Metric {self.monitor} improved by {abs(self.best_score - current):.3f} >="
                    f" min_delta = {abs(self.min_delta)}. New best score: {current:.3f}"
                )
            else:  # self.threshold_mode == 'rel':
                msg = (
                    f"Metric {self.monitor} improved by {abs(self.best_score - current) / self.best_score:.3f} >="
                    f" min_delta = {abs(self.min_delta)}. New best score: {current:.3f}"
                ) </b>
        else:
            msg = f"Metric {self.monitor} improved. New best score: {current:.3f}"
        return msg
</pre>

If you find it useful, I could open a PR to add this feature.

**Comment by jogepari** [comment]
I'd like to see this included in lightning as well.

Thanks to @BrunoBelucci for his version. The relative case didn't quite work for me in mode "min" (turns into `inf - inf`)

I propose condition looking like this:

```python
elif (self.threshold_mode == 'rel'
              and self.monitor_op(current, (self.best_score * 
                                            (1 + self.min_delta * torch.sign(self.best_score))).to(current.device))):
```

<details>
  <summary>Full class with this fix and minor improvements:</summary>

```python
class EarlyStopping_mod(pl.callbacks.EarlyStopping):
    def __init__(self, *args, threshold_mode: Literal['abs', 'rel'] = 'abs', **kwargs):
        self.threshold_mode = threshold_mode
        super().__init__(*args, **kwargs)

    def _evaluate_stopping_criteria(self, current: Tensor) -> Tuple[bool, Optional[str]]:
        should_stop = False
        reason = None
        if self.check_finite and not torch.isfinite(current):
            should_stop = True
            reason = (
                f"Monitored metric {self.monitor} = {current} is not finite."
                f" Previous best value was {self.best_score:.3f}. Signaling Trainer to stop."
            )
        elif self.stopping_threshold is not None and self.monitor_op(current, self.stopping_threshold):
            should_stop = True
            reason = (
                "Stopping threshold reached:"
                f" {self.monitor} = {current} {self.order_dict[self.mode]} {self.stopping_threshold}."
                " Signaling Trainer to stop."
            )
        elif self.divergence_threshold is not None and self.monitor_op(-current, -self.divergence_threshold):
            should_stop = True
            reason = (
                "Divergence threshold reached:"
                f" {self.monitor} = {current} {self.order_dict[self.mode]} {self.divergence_threshold}."
                " Signaling Trainer to stop."
            )
        elif (self.threshold_mode == 'abs'
              and self.monitor_op(current - self.min_delta, self.best_score.to(current.device))):
            should_stop = False
            reason = self._improvement_message(current)
            self.best_score = current
            self.wait_count = 0
        elif (self.threshold_mode == 'rel'
              and self.monitor_op(current, (self.best_score * 
                                            (1 + self.min_delta * torch.sign(self.best_score))).to(current.device))):
            should_stop = False
            reason = self._improvement_message(current)
            self.best_score = current
            self.wait_count = 0
        else:
            self.wait_count += 1
            if self.wait_count >= self.patience:
                should_stop = True
                reason = (
                    f"Monitored metric {self.monitor} did not improve in the last {self.wait_count} records."
                    f" Best score: {self.best_score:.3f}. Signaling Trainer to stop."
                )

        return should_stop, reason

    def _improvement_message(self, current: Tensor) -> str:
        """Formats a log message that informs the user about an improvement in the monitored score."""
        if torch.isfinite(self.best_score):
            if self.threshold_mode == 'abs':
                msg = (
                    f"Metric {self.monitor} improved by {abs(self.best_score - current):.3f} >="
                    f" min_delta = {abs(self.min_delta)}. New best score: {current:.3f}"
                )
            else:  # self.threshold_mode == 'rel':
                msg = (
                    f"Metric {self.monitor} improved by {abs(self.best_score - current) / abs(self.best_score):.3f} >="
                    f" min_delta = {abs(self.min_delta)}. New best score: {current:.3f}"
                ) 
        else:
            msg = f"Metric {self.monitor} improved. New best score: {current:.3f}"
        return msg
```
  
</details>

**Comment by mazurkin** [comment]
I just created a new class and reused `self.min_delta` which became the relative coefficient instead of absolute delta:

```python
import typing as t
import torch
import lightning.pytorch as pl

class RelativeEarlyStopCallback(pl.callbacks.EarlyStopping):
    """Early stopping callback that uses relative (proportional) delta instead of absolute delta.

    The standard ``EarlyStopping`` uses an absolute ``min_delta`` threshold: the metric must
    improve by at least ``min_delta`` units to count as an improvement. This works poorly when
    the metric magnitude varies across training runs or stages — a fixed absolute delta can be
    too lenient for large values and too strict for small values.

    ``RelativeEarlyStopCallback`` reinterprets ``min_delta`` as a *relative fraction* of the current
    metric value. For example, ``min_delta=0.01`` requires a 1% improvement relative to the
    current score, regardless of its magnitude.

    The improvement check comparison:
    - Original (absolute):  ``monitor_op(current - min_delta, best_score)``
    - Relative (this class): ``monitor_op(current + |current| * min_delta, best_score)``

    Note: ``min_delta`` is sign-adjusted in the parent ``__init__`` (positive for ``max`` mode,
    negative for ``min`` mode), so the formula works correctly for both modes.

    All other stopping criteria (non-finite check, stopping threshold, divergence threshold,
    patience) are identical to the parent class.
    """

    @t.override
    def _evaluate_stopping_criteria(self, current: torch.Tensor) -> tuple[bool, t.Optional[str]]:
        """Evaluate whether training should stop based on the current metric value.

        Overrides the parent to use a relative delta: the metric must improve by at least
        ``min_delta`` fraction of its current absolute value to reset the patience counter.

        Args:
            current: the current value of the monitored metric

        Returns:
            a tuple of (should_stop, reason) where reason is a human-readable message or None
        """
        should_stop: bool = False
        reason: str | None = None

        if self.check_finite and not torch.isfinite(current):
            should_stop = True
            self.stopping_reason = lightning.pytorch.callbacks.early_stopping.EarlyStoppingReason.NON_FINITE_METRIC
            reason = (
                f'Monitored metric {self.monitor} = {current} is not finite.'
                f' Previous best value was {self.best_score:.3f}. Signaling Trainer to stop.'
            )
        elif self.stopping_threshold is not None and self.monitor_op(current, self.stopping_threshold):
            should_stop = True
            self.stopping_reason = lightning.pytorch.callbacks.early_stopping.EarlyStoppingReason.STOPPING_THRESHOLD
            reason = (
                'Stopping threshold reached:'
                f' {self.monitor} = {current} {self.order_dict[self.mode]} {self.stopping_threshold}.'
                ' Signaling Trainer to stop.'
            )
        elif self.divergence_threshold is not None and self.monitor_op(-current, -self.divergence_threshold):
            should_stop = True
            self.stopping_reason = lightning.pytorch.callbacks.early_stopping.EarlyStoppingReason.DIVERGENCE_THRESHOLD
            reason = (
                'Divergence threshold reached:'
                f' {self.monitor} = {current} {self.order_dict[self.mode]} {self.divergence_threshold}.'
                ' Signaling Trainer to stop.'
            )
        else:
            # here we treat `self.min_delta` as a relative delta coefficient,
            # not the absolute delta as in the original implementation

            # move the best score to the `current` tensor's device to be compared later
            best_score: torch.Tensor = self.best_score.to(current.device)

            # raise the bar for `max` mode (delta > 0) or lower it for `min` mode (delta < 0),
            # so `current` must beat `best_score` by at least `min_delta` fraction to count
            # delta `self.min_delta` is already multiplied by `-1` in constructor for the `min` mode
            best_score_with_delta: torch.Tensor = best_score + best_score.abs() * self.min_delta

            improvement: bool = self.monitor_op(current, best_score_with_delta)

            if improvement:
                should_stop = False
                reason = self._improvement_message(current)
                self.best_score = current
                self.wait_count = 0
            else:
                self.wait_count += 1
                if self.wait_count >= self.patience:
                    should_stop = True
                    self.stopping_reason = (
                        lightning.pytorch.callbacks.early_stopping.EarlyStoppingReason.PATIENCE_EXHAUSTED
                    )
                    reason = (
                        f'Monitored metric {self.monitor} did not improve'
                        f' in the last {self.wait_count} records.'
                        f' Best score: {self.best_score:.3f}. Signaling Trainer to stop.'
                    )

        return should_stop, reason

```
